Add inline fast paths for SymInt operators

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #161586
* #161466

If SymInt::maybe_as_int() returns non-empty, then we get an inline
fast path. The philosophy here (as with the previous PR) is to
preserve performance in the "plain old ints" case.

Observed time spent in SymInt functions in computeStorageNBytes to
drop (and not cost shift elsewhere in the function) after this change,
profiling detach() using code similar to the benchmark from #160580
and Linux perf.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

Differential Revision: [D81530107](https://our.internmc.facebook.com/intern/diff/D81530107)

swolchok committed:
Add inline fast paths for SymInt operators

If SymInt::maybe_as_int() returns non-empty, then we get an inline
fast path. The philosophy here (as with the previous PR) is to
preserve performance in the "plain old ints" case.

Observed time spent in SymInt functions in computeStorageNBytes to
drop (and not cost shift elsewhere in the function) after this change,
profiling detach() using code similar to the benchmark from #160580
and Linux perf.

[ghstack-poisoned]

ezyang commented on c10/test/core/SymInt_test.cpp:
```diff
@@ -35,4 +36,167 @@ TEST(SymIntTest, Overflows) {
 
+namespace {
+
+class ConstantIntPretendingToBeSymbolicSymNodeImpl
```
What's going on here?

swolchok replied:
I needed a SymNodeImpl that would actually hit the `!maybe_as_int()` path in all the operators (and have working arithmetic) so I could get full test coverage. `ConstantSymNodeImpl` is `maybe_as_int()`. 

I'll add a code comment for this.

swolchok committed:
Update on "Add inline fast paths for SymInt operators"


If SymInt::maybe_as_int() returns non-empty, then we get an inline
fast path. The philosophy here (as with the previous PR) is to
preserve performance in the "plain old ints" case.

Observed time spent in SymInt functions in computeStorageNBytes to
drop (and not cost shift elsewhere in the function) after this change,
profiling detach() using code similar to the benchmark from #160580
and Linux perf.

[ghstack-poisoned]

swolchok committed:
rebase on "Add inline fast paths for SymInt operators"


If SymInt::maybe_as_int() returns non-empty, then we get an inline
fast path. The philosophy here (as with the previous PR) is to
preserve performance in the "plain old ints" case.

Observed time spent in SymInt functions in computeStorageNBytes to
drop (and not cost shift elsewhere in the function) after this change,
profiling detach() using code similar to the benchmark from #160580
and Linux perf.

[ghstack-poisoned]

swolchok committed:
Update on "Add inline fast paths for SymInt operators"


If SymInt::maybe_as_int() returns non-empty, then we get an inline
fast path. The philosophy here (as with the previous PR) is to
preserve performance in the "plain old ints" case.

Observed time spent in SymInt functions in computeStorageNBytes to
drop (and not cost shift elsewhere in the function) after this change,
profiling detach() using code similar to the benchmark from #160580
and Linux perf.

[ghstack-poisoned]

swolchok commented:
@swolchok has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D81530107).

swolchok committed:
rebase on "Add inline fast paths for SymInt operators"


If SymInt::maybe_as_int() returns non-empty, then we get an inline
fast path. The philosophy here (as with the previous PR) is to
preserve performance in the "plain old ints" case.

Observed time spent in SymInt functions in computeStorageNBytes to
drop (and not cost shift elsewhere in the function) after this change,
profiling detach() using code similar to the benchmark from #160580
and Linux perf.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

Differential Revision: [D81530107](https://our.internmc.facebook.com/intern/diff/D81530107)

[ghstack-poisoned]

ezyang reviewed (approved):
skip sounds good

swolchok commented:
@swolchok has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D81530107).

swolchok commented:
@pytorchbot merge

swolchok commented:
@pytorchbot merge

swolchok commented:
HUD says the failling inductor jobs were broken today, and I had rebased this morning. ignoring current failures (inductor / inductor-build, inductor / unit-test / inductor-halid-build, and and inductor / unit-test / inductor-triton-cpu-build) accordingly.
